### PR TITLE
Pass the type information from the parser to the instruction

### PR DESCRIPTION
**Pull request by fcastellacci**
I had to comment out an Assert to make everything work. Need to find a proper work-around

### CODE REVIEW

File: lib/Backend/IRBuilder.cpp [hunk 1]
```diff
@@ -4501,6 +4505,32 @@ IRBuilder::BuildElementCP(Js::OpCode newOpcode, uint32 offset, Js::RegSlot insta
...
+            switch (((Js::FunctionBody::TypeInformation*)typeAnnotation)->type)
+            {
+            case Js::TypeHint::Int:
+                instr->AsProfiledInstr()->u.FldInfo().valueType = ValueType::Int.SetCanBeTaggedValue(true);
+                break;
+            case Js::TypeHint::Float:
+                instr->AsProfiledInstr()->u.FldInfo().valueType = ValueType::Float.SetCanBeTaggedValue(true);
+                break;
+            case Js::TypeHint::Bool:
+                instr->AsProfiledInstr()->u.FldInfo().valueType = ValueType::Boolean.SetCanBeTaggedValue(true);
```
Comment (MikeHolman) [inline lib/Backend/IRBuilder.cpp]: I'm not sure this will work. I think we store booleans as Objects

File: lib/Backend/IRBuilder.cpp [hunk 2]
```diff
@@ -4501,6 +4505,32 @@ IRBuilder::BuildElementCP(Js::OpCode newOpcode, uint32 offset, Js::RegSlot insta
...
+            m_func->GetJITFunctionBody()->GetTypeAnnotationsArray()->count,
+            sizeof(Js::FunctionBody::TypeInformation), compareTypeInfos)) != nullptr)
+        {
+            switch (((Js::FunctionBody::TypeInformation*)typeAnnotation)->type)
+            {
+            case Js::TypeHint::Int:
+                instr->AsProfiledInstr()->u.FldInfo().valueType = ValueType::Int.SetCanBeTaggedValue(true);
+                break;
+            case Js::TypeHint::Float:
+                instr->AsProfiledInstr()->u.FldInfo().valueType = ValueType::Float.SetCanBeTaggedValue(true);
```
Comment (MikeHolman) [inline lib/Backend/IRBuilder.cpp]: i think for x86, CanBeTagged should be false

File: lib/JITIDL/JITTypes.h [hunk 3]
```diff
@@ -466,6 +466,19 @@ typedef struct ConstTableContentIDL
...
+typedef struct TypeInformationIDL {
+    unsigned int bytecodeOffset;
+    unsigned int regSlot;
+    unsigned char type;
+} TypeInformationIDL;
+
+typedef struct TypeAnnotationsArrayIDL
+{
+    unsigned int count;
+    //Should I use some padding?
```
Comment (MikeHolman) [inline lib/JITIDL/JITTypes.h]: yeah, but don't worry about it right now. that is just for perf for marshalling

File: lib/Parser/Parse.cpp [hunk 4]
```diff
@@ -6307,7 +6314,33 @@ void Parser::ParseFncFormals(ParseNodePtr pnodeFnc, ParseNodePtr pnodeParentFnc,
                 }
 
                 m_pscan->Scan();
+                //FCASTE: Process type annotation on param here
+                //FCASTE: after parsing a term, check if there is a type annotation and parse it
+                if (CONFIG_FLAG(TypeAnnotations) && m_token.tk == tkTypeAnnBegin)
```
Comment (MikeHolman) [inline lib/Parser/Parse.cpp]: can you move this out to its own method to avoid copy/paste?

**Comment by MikeHolman** [comment]
:shipit:

File: lib/Backend/IRBuilder.cpp (hunk 1 above)
Comment (fcastellacci) [inline lib/Backend/IRBuilder.cpp]: Yeah, you're right. Looks like booleans can't be set as tagged values.
I'll change that

File: lib/Backend/IRBuilder.cpp (hunk 2 above)
Comment (fcastellacci) [inline lib/Backend/IRBuilder.cpp]: I tested it with annotated js in x86 and it didn't break. 
Should I change it anyway?

File: lib/Parser/Parse.cpp (hunk 4 above)
Comment (fcastellacci) [inline lib/Parser/Parse.cpp]: Done

**Pull request merged**